src/transform/src/redundant_join.rs

```diff
@@ -122,26 +130,34 @@ impl RedundantJoin {
                     // We descend into the values and the body with the naively
                     // extended context.
                     for value in values.iter_mut() {
-                        self.action(value, lets)?;
+                        self.action(value, ctx)?;
+                    }
+                    // Clear uses from the just visited recursive binding
+                    // definitions.
+                    for value in values.iter_mut() {
+                        ctx.remove_uses(value);
                     }
-                    let result = self.action(body, lets)?;
+                    let result = self.action(body, ctx)?;
 
                     // Remove the lets entries for all ids.
                     for id in ids.iter() {
-                        lets.remove(&Id::Local(*id));
+                        ctx.remove(id);
                     }
 
                     Ok(result)
                 }
 
                 MirRelationExpr::Get { id, typ } => {
-                    // Extract the value provenance, or an empty list if
-                    // unavailable (this should only be the case for GlobalId
-                    // references).
-                    let mut val_info = lets.get(id).cloned().unwrap_or_default();
-                    // Add information about being exactly this let binding too.
-                    val_info.push(ProvInfo::make_leaf(*id, typ.arity()));
-                    Ok(val_info)
+                    if let Id::Local(id) = id {
+                        // Extract the value provenance (this should always exist).
```

Comment:
If this should always exist, wouldn't it be better to use `expect`?

Reply:
In theory yes - in practice I'm not 100% sure that this won't lead to new panics in production because the line of reasoning that asserts this is kind of tricky. TL;DR:

1. We count all uses only once in `transform` before we call `action` with the root of the given `relation`.
2. We remove uses in the `Let` / `LetRec` / `Join` cases of the `action`.
3. Whenever a `ctx.uses` entry resets to zero, we remove the associated `ctx.lets` entry.
4. However, since `action` also occasionally modifies the tree (in `join`) in theory it could happen that our counts no longer reflect reality.

With the current version we will fall back to a more conservative approach if for some reason the expected invariant is violated. I'm happy to rewrite this as a `soft_panic_or_assert!` for now, and then maybe promote to `assert` after we get some more confidence.